huaweicloud/resource_huaweicloud_bms_instance.go

```diff
@@ -0,0 +1,589 @@
+		}
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"Deleting", "ACTIVE", "SHUTOFF"},
+		Target:       []string{"DELETED"},
+		Refresh:      waitForBmsInstanceDelete(bmsClient, d.Id()),
+		Timeout:      d.Timeout(schema.TimeoutDelete),
+		Delay:        60 * time.Second,
+		PollInterval: 20 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error deleting HuaweiCloud BMS instance: %s", err)
+	}
+
+	d.SetId("")
+	return nil
+}
+func resourceBmsInstanceNicsV1(d *schema.ResourceData) []baremetalservers.Nic {
+	var nicRequests []baremetalservers.Nic
+
+	nics := d.Get("nics").([]interface{})
+	for i := range nics {
+		nic := nics[i].(map[string]interface{})
+		nicRequest := baremetalservers.Nic{
+			SubnetId:  nic["subnet_id"].(string),
+			IpAddress: nic["ip_address"].(string),
+		}
+
+		nicRequests = append(nicRequests, nicRequest)
+	}
+	return nicRequests
+}
+
+func resourceBmsInstanceDataVolumesV1(d *schema.ResourceData) []baremetalservers.DataVolume {
+	var volRequests []baremetalservers.DataVolume
+
+	vols := d.Get("data_disks").([]interface{})
+	for i := range vols {
+		vol := vols[i].(map[string]interface{})
+		volRequest := baremetalservers.DataVolume{
+			VolumeType: vol["type"].(string),
+			Size:       vol["size"].(int),
+		}
+		volRequests = append(volRequests, volRequest)
+	}
+	return volRequests
+}
+
+func resourceBmsInstanceSecGroupsV1(d *schema.ResourceData) []baremetalservers.SecurityGroup {
+	rawSecGroups := d.Get("security_groups").(*schema.Set).List()
+	secgroups := make([]baremetalservers.SecurityGroup, len(rawSecGroups))
+	for i, raw := range rawSecGroups {
+		secgroups[i] = baremetalservers.SecurityGroup{
+			ID: raw.(string),
+		}
+	}
+	return secgroups
+}
+
+func flattenBmsInstanceNicsV1(d *schema.ResourceData, meta interface{},
+	addresses map[string][]baremetalservers.Address) []map[string]interface{} {
+
+	config := meta.(*config.Config)
+	networkingClient, err := config.NetworkingV2Client(GetRegion(d, config))
+	if err != nil {
+		log.Printf("Error creating HuaweiCloud networking client: %s", err)
+	}
+
+	var network string
+	nics := []map[string]interface{}{}
+	// Loop through all networks and addresses.
+	for _, addrs := range addresses {
+		for _, addr := range addrs {
+			// Skip if not fixed ip
+			if addr.Type != "fixed" {
+				continue
+			}
+
+			p, err := ports.Get(networkingClient, addr.PortID).Extract()
+			if err != nil {
+				network = ""
+				log.Printf("[DEBUG] flattenInstanceNicsV1: failed to fetch port %s", addr.PortID)
+			} else {
+				network = p.NetworkID
+			}
+
+			v := map[string]interface{}{
+				"subnet_id":   network,
+				"ip_address":  addr.Addr,
+				"mac_address": addr.MacAddr,
+				"port_id":     addr.PortID,
+			}
+			nics = append(nics, v)
+		}
+	}
+
+	log.Printf("[DEBUG] flattenInstanceNicsV1: %#v", nics)
+	return nics
+}
+
+func bmsPublicIP(server *baremetalservers.CloudServer) string {
```

Comment:
can we use the similar function in `huaweicloud_compute_instance`?

Reply:
fixed

Reply:
the type of param `server` is different